pkg/dns/dns.go

```diff
@@ -0,0 +1,394 @@
+	if err != nil {
+		return nil, err
+	}
+	if dnsConfig != nil {
+		for _, s := range dnsConfig.Servers {
+			r.resolvConfServers = append(r.resolvConfServers, net.JoinHostPort(s, dnsConfig.Port))
+		}
+	}
+
+	return r, nil
+}
+
+func (r *DNSResolver) StartDNSResolver(stopCh <-chan struct{}) {
+	go r.startResolver()
+	go r.refreshDNS()
+	go func() {
+		<-stopCh
+		r.dnsRefreshQueue.ShutDown()
+		close(r.DnsResolverChan)
+	}()
+}
+
+// startResolver watches the DnsResolver Channel
+func (r *DNSResolver) startResolver() {
+	temp := make(chan struct{}, MaxConcurrency)
+	for clusters := range r.DnsResolverChan {
+		temp <- struct{}{}
+
+		go func(clusters []*config_cluster_v3.Cluster) {
+			defer func() { <-temp }()
+			r.resolveDomains(clusters)
+		}(clusters)
+	}
+}
+
+// resolveDomains takes a map of hostnames and refresh rate
+func (r *DNSResolver) resolveDomains(clusters []*config_cluster_v3.Cluster) {
+	domains := getPendingResolveDomain(clusters)
+
+	// Stow domain updates, need to remove unwatched domains first
+	r.removeUnwatchedDomain(domains)
+	for _, v := range domains {
+		r.resolve(v)
+	}
+	r.adsCache.ClusterCache.Flush()
+}
+
+// removeUnwatchedDomain cancels any scheduled re-resolve for names we no longer care about
+func (r *DNSResolver) removeUnwatchedDomain(domains map[string]*pendingResolveDomain) {
+	r.Lock()
+	defer r.Unlock()
+	for name, value := range r.cache {
+		if _, ok := domains[name]; ok {
+			continue
+		}
+		r.adsCache.UpdateApiClusterStatus(value.clusterName, core_v2.ApiStatus_DELETE)
+		delete(r.cache, name)
+	}
+}
+
+// This functions were copied and adapted from github.com/istio/istio/pilot/pkg/model/network.go.
+func (r *DNSResolver) resolve(v *pendingResolveDomain) {
+	if entry, ok := r.cache[v.domainName]; ok && entry.expiry.After(time.Now()) {
+		return
+	}
+
+	r.Lock()
+	// ideally this will not happen more than once for each name and the cache auto-updates in the background
+	// even if it does, this happens on the SotW ingestion path (kube or meshnetworks changes)
+	entry := r.cache[v.domainName]
+	delete(r.cache, v.domainName)
+	r.Unlock()
+
+	addrs, ttl, err := r.doResolve(v.domainName, v.refreshRate)
+	// refresh the dns address periodically by respecting the dnsRefreshRate and ttl, which one is shorter
+	if ttl > v.refreshRate {
+		ttl = v.refreshRate
+	}
+	expiry := time.Now().Add(ttl)
+	if err != nil {
+		// gracefully retain old addresses in case the DNS server is unavailable
+		addrs = entry.value
+	}
+
+	r.cache[v.domainName] = &domainCacheEntry{
+		value:       addrs,
+		clusterName: v.cluster.GetName(),
+		expiry:      expiry,
+	}
+
+	// push to refresh queue
+	r.dnsRefreshQueue.AddAfter(v, time.Until(expiry))
+
+	if entry == nil {
+		// for the newly resolved domain just push to bpf map
+		log.Infof("resolve dns , name: %s, addr: %v\n", v.domainName, addrs)
+		v.setAddrsToCluster(addrs)
+		r.adsCache.CreateApiClusterByCds(core_v2.ApiStatus_UPDATE, v.cluster)
```

Comment:
Think of a case, during the resolve, what if a cluster is deleted. Do you have a chance to delete it in the cache